### PR TITLE
Initial version

### PR DESCRIPTION
**Pull request by bojidar-bg**
Depends on https://github.com/comrade-coop/strongforce/pull/18 being merged and released.

Will fix tests failing after that.

### CODE REVIEW

**Review by ivanasen — APPROVED** [review]

File: src/Achievements/Achievement.cs [hunk 1]
```diff
@@ -19,14 +21,42 @@ public class Achievement : UniformTokenSplitter
...
+		}
 		public override IDictionary<string, object> GetState()
 		{
 			var state = base.GetState();
 
 			state.Set("MintTokenManager", this.MintTokenManager);
 			state.Set("BurnTokenManager", this.BurnTokenManager);
 			state.Set("Exchanger", this.Exchanger);
-
+			state.Set("Describtion", this.Describtion);
```
Comment (bojidar-bg) [inline src/Achievements/Achievement.cs]: @IoanStoianov "Descri**b**tion"?

File: src/Achievements/Achievement.cs [hunk 2]
```diff
@@ -19,14 +21,42 @@ public class Achievement : UniformTokenSplitter
...
+
+		protected IDictionary<Address, ITaggedTokens> TokenContributors { get; set; }
+			= new Dictionary<Address, ITaggedTokens>();
+
+		protected override void HandleMessage(Message message)
+		{
+			switch (message.Type)
+			{
+				case TokensReceivedEvent.Type:
+					var sender = message.Payload.Get<Address>(TokensReceivedEvent.From);
```
Comment (ivanasen) [inline src/Achievements/Achievement.cs]: Use explicit types

File: src/Achievements/Achievement.cs [hunk 3]
```diff
@@ -19,14 +21,42 @@ public class Achievement : UniformTokenSplitter
...
+					var tokens = message.Payload.GetDictionary(TokensReceivedEvent.TokensTotal);
+					if(sender != null)
+					{
+						if(this.TokenContributors.ContainsKey(sender)) this.TokenContributors[sender].AddToBalance(new TaggedTokens(tokens));
```
Comment (ivanasen) [inline src/Achievements/Achievement.cs]: Too long line. Use auto-formatting

File: src/Achievements/Achievement.cs [hunk 4]
```diff
@@ -19,14 +21,42 @@ public class Achievement : UniformTokenSplitter
...
+						else this.TokenContributors.Add(sender, new TaggedTokens(tokens));
+					}
+						
+					this.Split(message.Sender, new ReadOnlyTaggedTokens(tokens));
+					break;
+				default:
+					base.HandleMessage(message);
+					return;
+			}
+		}
```
Comment (ivanasen) [inline src/Achievements/Achievement.cs]: Missing new line

File: src/Achievements/Achievement.cs [hunk 5]
```diff
@@ -19,14 +21,42 @@ public class Achievement : UniformTokenSplitter
...
+					{
+						if(this.TokenContributors.ContainsKey(sender)) this.TokenContributors[sender].AddToBalance(new TaggedTokens(tokens));
+						else this.TokenContributors.Add(sender, new TaggedTokens(tokens));
+					}
+						
+					this.Split(message.Sender, new ReadOnlyTaggedTokens(tokens));
+					break;
+				default:
+					base.HandleMessage(message);
+					return;
```
Comment (ivanasen) [inline src/Achievements/Achievement.cs]: Unnecessary return statement

File: src/Achievements/AchievementsKit.cs [hunk 6]
```diff
@@ -53,6 +55,7 @@ protected override void Instantiate(Address initialManager)
 			{
 				{ "Admin", initialManager.ToString() },
 				{ "User", AccessControlList.AnyAddress },
+				{ "Achievements", achievements },
```
Comment (ivanasen) [inline src/Achievements/AchievementsKit.cs]: Do we really need to pass in an empty list? We can just make the parameter optional

File: src/Achievements/Achievements.csproj [hunk 7]
```diff
@@ -1,12 +1,11 @@
...
-  <PropertyGroup>
-    <TargetFramework>netcoreapp2.2</TargetFramework>
-  </PropertyGroup>
+<Project Sdk="Microsoft.NET.Sdk">
 
+  <PropertyGroup>
+    <TargetFramework>netcoreapp2.2</TargetFramework>
+  </PropertyGroup>
   <ItemGroup>
-    <PackageReference Include="Wetonomy.TokenSystem" Version="0.0.0" />
```
Comment (ivanasen) [inline src/Achievements/Achievements.csproj]: We should maybe revert these. To satisfy the dependencies we can use local Nuget packages.

File: src/Achievements/AchievementFactory.cs [hunk 8]
```diff
@@ -55,29 +60,35 @@ protected override void HandleMessage(Message message)
...
 		{
+			var recipients = payload.GetList<Address>("Recipients").ToHashSet();
+			var describtion = payload.Get<string>("Describtion");
 			var address = this.CreateContract<Achievement>(new Dictionary<string, object>()
 			{
 				{ "MintTokenManager", this.MintTokenManager.ToString() },
 				{ "BurnTokenManager", this.BurnTokenManager.ToString() },
 				{ "Exchanger", this.Address.ToString() },
+				{ "Describtion", describtion },
+				// { "TokenContributors", new Dictionary<Address, object>()}
```
Comment (ivanasen) [inline src/Achievements/AchievementFactory.cs]: Remove commented line

**Review by ivanasen — CHANGES_REQUESTED** [review]

File: src/Achievements/Achievement.cs (hunk 2 above)
Comment (bojidar-bg) [reply src/Achievements/Achievement.cs]: The type (`Address`) is already present on the line.

File: src/Achievements/Achievement.cs (hunk 5 above)
Comment (bojidar-bg) [reply src/Achievements/Achievement.cs]: Nope. It won't compile without a `break` or `return` there.

**Pull request merged**